impl/src/main/java/org/glassfish/soteria/authorization/spi/impl/SubjectParser.java

```diff
@@ -389,181 +383,198 @@ private boolean tryWebLogic(String contextID, Collection<String> allDeclaredRole
         List<String> groups = new ArrayList<>();
         principalToGroups(principal, groups);
         return groups;
     }
-    
+
     private Principal doGetCallerPrincipalFromPrincipals(Iterable<Principal> principals) {
-        
-        for (Principal principal : principals) {
-            switch (principal.getClass().getName()) {
-                case "org.glassfish.security.common.PrincipalImpl": // GlassFish/Payara
-                // JBoss EAP/WildFly convention 1 - single top level principal of the below type
-                case "org.jboss.security.SimplePrincipal":
-                case "com.ibm.ws.security.authentication.principals.WSPrincipal": // Liberty
-                    return principal;
-                // JBoss EAP/WildFly convention 2 - the one and only principal in group called CallerPrincipal
-                case "org.jboss.security.SimpleGroup": 
-                    if (principal.getName().equals("CallerPrincipal") && principal instanceof Group) {
-                        
-                        Enumeration<? extends Principal> groupMembers = ((Group) principal).members();
-                        
-                        if (groupMembers.hasMoreElements()) {
-                            return groupMembers.nextElement();
-                        }
-                    }
-                    break;
-                case "org.apache.tomee.catalina.TomcatSecurityService$TomcatUser": // TomEE
-                    try {
-                        return 
-                            (Principal) Class.forName("org.apache.catalina.realm.GenericPrincipal")
-                                             .getMethod("getUserPrincipal")
-                                             .invoke(
-                                                 Class.forName("org.apache.tomee.catalina.TomcatSecurityService$TomcatUser")
-                                                      .getMethod("getTomcatPrincipal")
-                                                      .invoke(principal))
-                            
-                            ;
-                        } catch (Exception e) {
-                            
-                        }
-                    break;
-            }
+        // Check for Servlet
+        try {
+            return CDI.current().select(HttpServletRequest.class).get().getUserPrincipal();
+        } catch (Exception e) {
+            // Not inside an HttpServletRequest
+        }
+
+        // Check for EJB
+        EJBContext ejbContext = EJB.getEJBContext();
+        if (ejbContext != null) {
+            // The EJB returned value must be verified for its "unauthenticated name" since it's vendor specific
+            return getVendorCallerPrincipal(ejbContext.getCallerPrincipal(), true);
+        }
 
+        for (Principal principal : principals) {
+            // If the principal is a standard one, assume it's the "application" caller principal and return it
```

Comment:
The application principal should never be returned here; this is the platform-specific caller principal. The platform caller principal could be represented by the type `CallerPrincipal`, but this check should probably be the last one, after all the proprietary principal types are checked, since an actual `CallerPrincipal` is probably the least likely representation at this point. It could even be done inside `getVendorCallerPrincipal()`, as the last check.